### ISSUE TITLE
Create a Reactive Stream as an extension

### ISSUE THREAD

**Issue by otaviojava** [eclipse/jnosql#232]
## Goal 

The goal of this proposal is to write an extension to each API that allows integration with Reactive Streams API.
Briefly, Reactive Stream is an initiative to provide a standard for asynchronous stream processing with non-blocking backpressure.

This extension will work integrated with the [Eclipse MicroProfile Reactive Streams Operators](https://github.com/eclipse/microprofile-reactive-streams-operators/blob/master/README.adoc).

### About Reactive Streams

Reactive Streams is an integration SPI - it allows two different libraries that provide asynchronous streaming to be able to stream data to and from each other.

Reactive Streams is not however designed to be used directly by application developers. The semantics defined by Reactive Streams are very strict, and are nontrivial, particularly in the areas of thread safety, to implement correctly. Typically, application developers are expected to use third-party libraries that provide the tools necessary to manipulate and control streams. Examples include [Akka Streams](https://doc.akka.io/docs/akka/current/stream/index.html), [RxJava](https://github.com/ReactiveX/RxJava), and [Reactor](https://projectreactor.io/).

### Repository

The repositories programming model is the most high-level abstraction and we'll have a repository to handle with `ReactiveRepository`:

```java

public interface GodRepository extends ReactiveRepository<God, String> {
}
```

```java

String id =...;
Publisher<God> publisher = repository.findById("id");
CompletionSubscriber<God, Optional<God>> subscriber = ReactiveStreams.<God>builder().findFirst().build();
publisher.subscribe(subscriber);
CompletionStage<Optional<God>> completion = subscriber.getCompletion();
```

### Templates

Just as the traditional repositories are based on traditional template implementations, the reactive ones are built on top of a reactive template.

* ReactiveKeyValueTemplate
* ReactiveDocumentTemplate
* ReactiveColumnTemplate

```java
ReactiveKeyValueTemplate template
Publisher<God> publisher = template.get("id", God.class);
CompletionSubscriber<God, Optional<God>> subscriber = ReactiveStreams.<God>builder().findFirst().build();
publisher.subscribe(subscriber);
CompletionStage<Optional<God>> completion = subscriber.getCompletion();
```
References: 
* https://spring.io/blog/2016/11/28/going-reactive-with-spring-data
* [MicroProfile Reactive Streams Operators Specification](https://download.eclipse.org/microprofile/microprofile-reactive-streams-operators-1.0/microprofile-reactive-streams-operators-spec.pdf)

**Comment by keilw** [comment]
This is a similar situation as #203. While it seems fine for JNoSQL to use a MicroProfile library in an implementation, it is not acceptable for the Spec/API (Jakarta NoSQL)

**Comment by hantsy** [comment]
And do not forget the Java 9 Flow API, it should be supported, like Spring Data,  as an addon of the ReactiveStreams spec can be converted between different ReactiveStreams impls.

**Comment by Emily-Jiang** [comment]
@otaviojava Did you look at https://spring.io/projects/spring-data-r2dbc? As for which API you are going to support, I think supporting Java 9 Flow is better than org.reactivestreams APIs. However, you can use the Reactive Streams RxJava, Reactor, Akka for implementations. Adopting MicroProfile Reactive Streams Operators is a great idea as well.

**Comment by otaviojava** [comment]
@Emily-Jiang 
Currently, Jakarta EE 8 has support for Java 8; therefore, we cannot use Java 9.
Hopefully, Reactive Streams Operators can have an implementation that has support to `Flow` in Java 9 or higher.

**Comment by hantsy** [comment]
@otaviojava I do not think it is conflict when adopting Java 9 Flow APIs.  We can handle it like Spring Data, support ReactiveStreams spec/Microprofile ReactiveStreams Operators which are Java 8 based, and add addons to support Java 9 Flow, Reactor, RxJava 1/2/3, SmallRye Munity, etc.

**Comment by otaviojava** [comment]
@hantsy That sounds like a good plan  :)

**Comment by keilw** [comment]
@otaviojava/all, please don't forget, that the final NoSQL release won't even be relevant to the Jakarta EE platform before Jakarta EE 10, so it is more than reasonable to consider adopting to Jakarta EE 9 now. Which is based on Java SE 9 or even 11, therefore all those reactive technologies could be fine, particularly for the implementation, as long as they are usable by Eclipse projects.

**Comment by otaviojava** [comment]
@keilw Eclipse MicroProfile uses interfaces, that mean we can use Flow.
Right now there three implementations on Eclipse MicroProfile:


* [Akka Streams](https://github.com/lightbend/microprofile-reactive-streams/tree/master/akka)
* [Zero Dependency ](https://github.com/lightbend/microprofile-reactive-streams/tree/master/zerodep)- intended as a possible reference implementation for when this is proposed to the JDK, but not the reference implementation for MicroProfile as MicroProfile does not have reference implementations.
* [SmallRye](https://github.com/smallrye/smallrye-reactive-streams-operators) - based on RxJava 2 and Eclipse Vert.x.

**Comment by otaviojava** [comment]
@Emily-Jiang 

What do you think about this Eclipse MicroProfile project?
Does it have a bright future?
Does it worth creating this kind of extension?

**Comment by Emily-Jiang** [comment]
@otaviojava I think it is a good idea to adopt MicroProfile Reactive Streams Operators even though in the future the APIs might change. I will be interested in any feedback you might have. 
By the way, I would not investigate any time to Zero Dependency as it is not maintained or used by any runtime. Focusing on SmallRye is the right thing to do.

**Comment by keilw** [comment]
@Emily-Jiang @otaviojava Note, that Jakarta EE has stronger compatibility requirements, therefore a frequently changing API is not so good. Let's wait till MicroProfile becomes more mature and especially for the API it can't be used until there is a resolution by the various committees following such committee or WG on the MP side, a current version of MP Reactive Streams can't be used. Isn't SmallRye just one implementation? Jakarta NoSQL and JNoSQL should work with all Jakarta EE compatible runtimes and implementations.

**Comment by otaviojava** [comment]
@keilw 
I agree with you Werner, that why we'll create it as an extension and not directly in the RI.
The goal to achieve here is an extension to get feedback from the community.

**Comment by keilw** [comment]
I think I would rather stick to something like https://github.com/reactive-streams/reactive-streams-jvm. "For this reason, the same interfaces provided by https://www.reactive-streams.org in the org.reactivestreams package are used instead." so MP Reactive does nothing more than use this library. If we want to use the nifty diagrams. then we may also consider using the MP project but similar to MP OpenAPI it is just a think layer or glue to another existing spec, so the spec or API it uses is of greater value here.

**Comment by otaviojava** [comment]
@keilw 
I've created the Proposal: 
https://github.com/eclipse/jnosql/pull/237
I created an `Observable` interface, therefore, we can change the implementation as soon we will change to Java 11.

Also, as you can see it is an extension; therefore, it is not on directly on the RI.

**Comment by keilw** [comment]
But what does that mean for the PR, would it work now or only after adopting Java 11?

**Comment by otaviojava** [comment]
@keilw it will work on both now and later with less impact to the users.

**Comment by hantsy** [comment]
Now Jakarta EE 10 is aligned to Java 11,  I think it is time to add ReactiveStreams. I think using Java 9 Flow API is OK.

(Java 11 HttpClient is a good example of using Flow API, eg. [HttpResponse.BodySubscriber](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpResponse.BodySubscriber.html)). 

There are some utilities in reactivestreams.org to convert bwteen Flow and ReactiveStreams.

**Comment by otaviojava** [comment]
We'll close this idea for now.
